### PR TITLE
Load ceiling

### PR DESCRIPTION
**Pull request by buckley-w-david**
This pull request adds functionality to transform and load ceiling data extracted in the extract component into mongodb

### CODE REVIEW

File: python/src/energuide/dwelling.py [hunk 1]
```diff
@@ -65,6 +65,56 @@ def from_data(cls, data: str) -> 'Region':
...
+
+        )
+
+    def to_dict(self):
+        return {
+            'label': self.label,
+            'typeEnglish': self.type_english,
+            'typeFrench': self.type_french,
+            'nominalRSI': self.nominal_rsi,
+            'nominalR': self.nominal_rsi*self._RSI_RATIO if self.nominal_rsi is not None else None,
```
Comment (JasonMWhite) [inline python/src/energuide/dwelling.py]: spaces around `*`

Comment (JasonMWhite) [reply python/src/energuide/dwelling.py]: and parentheses if necessary

File: python/src/energuide/dwelling.py [hunk 2]
```diff
@@ -65,6 +65,56 @@ def from_data(cls, data: str) -> 'Region':
...
+
+    area_metres: typing.Optional[float]
+    length_metres: typing.Optional[float]
+
+
+class Ceiling(_Ceiling):
+
+    _RSI_RATIO = 5.678263337
+    _FEET_SQUARED_RATIO = 10.7639
+    _FEET_RATIO = 3.28084
```
Comment (JasonMWhite) [inline python/src/energuide/dwelling.py]: these aren't really ratios. More like multipliers, or conversion factors.

File: python/src/energuide/dwelling.py [hunk 3]
```diff
@@ -89,25 +140,46 @@ class ParsedDwellingDataRow(_ParsedDwellingDataRow):
...
+                }
+            }
+        }
     }
 
     @classmethod
     def from_row(cls, row: reader.InputData) -> 'ParsedDwellingDataRow':
         validator = cerberus.Validator(cls._SCHEMA, allow_unknown=True)
         if not validator.validate(row):
+            # import pdb; pdb.set_trace()
```
Comment (JasonMWhite) [inline python/src/energuide/dwelling.py]: 🔥

File: python/src/energuide/dwelling.py [hunk 4]
```diff
@@ -89,25 +140,46 @@ class ParsedDwellingDataRow(_ParsedDwellingDataRow):
...
-            modification_date=parser.parse(row['MODIFICATIONDATE']),
-            year_built=row['YEARBUILT'],
-            city=row['CLIENTCITY'],
-            region=Region.from_data(row['HOUSEREGION']),
-            forward_sortation_area=row['forwardSortationArea']
+            eval_id=parsed['EVAL_ID'],
+            eval_type=EvaluationType.from_code(parsed['EVAL_TYPE']),
+            entry_date=parser.parse(parsed['ENTRYDATE']).date(),
+            creation_date=parser.parse(parsed['CREATIONDATE']),
+            modification_date=parser.parse(parsed['MODIFICATIONDATE']),
```
Comment (JasonMWhite) [inline python/src/energuide/dwelling.py]: Should we put `parser.parse` in the `coerce` method of the validator?

File: python/src/energuide/dwelling.py [hunk 5]
```diff
@@ -117,12 +189,14 @@ def __init__(self, *,
                  evaluation_type: EvaluationType,
                  entry_date: datetime.date,
                  creation_date: datetime.datetime,
-                 modification_date: datetime.datetime
+                 modification_date: datetime.datetime,
+                 ceilings: typing.List[Ceiling]
```
Comment (JasonMWhite) [inline python/src/energuide/dwelling.py]: add a `,` at the end to minimize line edits the next time we add something

File: python/tests/test_dwelling.py [hunk 6]
```diff
@@ -10,30 +10,52 @@
...
+            {
+                'label': 'Main attic',
+                'type_english': 'Attic/gable',
+                'type_french': 'Combles/pignon',
+                'nominal_rsi': '2.864',
+                'effective_rsi': '2.9463',
+                'area': '46.4515',
+                'length': '23.875',
+            }
+        ]
```
Comment (JasonMWhite) [inline python/tests/test_dwelling.py]: We should have `sample_input_e` start from `sample_input_d` and `merge` in a dict with the necessary changes.

File: python/tests/test_dwelling.py [hunk 7]
```diff
@@ -90,6 +112,25 @@ def test_from_code(self):
     def test_from_unknown_code(self):
         assert dwelling.Region.from_data('CA') == dwelling.Region.UNKNOWN
 
+class TestCeiling:
+
+    @pytest.fixture
+    def sample(self):
```
Comment (JasonMWhite) [inline python/tests/test_dwelling.py]: This same example is used in `sample_input_d` as well. Define in one place and use in both?

File: python/src/energuide/dwelling.py (hunk 4 above)
Comment (buckley-w-david) [reply python/src/energuide/dwelling.py]: Not unless we want to make a custom validator type. The coercion is applied before the validation, so if it coerced it into a type that isn't supported by default, we'd have to write a custom validation type for it, except if it didn't throw an exception when coercing, we automatically know it's valid, so it's a weird situation

File: python/src/energuide/dwelling.py [hunk 8]
```diff
@@ -65,6 +65,56 @@ def from_data(cls, data: str) -> 'Region':
...
 
+class _Ceiling(typing.NamedTuple):
+    label: typing.Optional[str]
+    type_english: typing.Optional[str]
+    type_french: typing.Optional[str]
+
+    nominal_rsi: typing.Optional[float]
+
+    effective_rsi: typing.Optional[float]
+
```
Comment (JasonMWhite) [inline python/src/energuide/dwelling.py]: 🔥  extra linebreaks

File: python/tests/test_dwelling.py [hunk 9]
```diff
@@ -6,35 +6,40 @@
...
-        'ENTRYDATE': '2018-02-01',
-        'CREATIONDATE': '2018-02-08 09:00:00',
-        'MODIFICATIONDATE': '2018-06-01 09:00:00',
-        'CLIENTCITY': 'Montreal',
-        'forwardSortationArea': 'G1A',
-        'HOUSEREGION': 'Quebec',
-        'YEARBUILT': 2001,
-    }
+def sample_input_e(sample_input_d: reader.InputData) -> reader.InputData:
+    sample_input_d.update(EVAL_TYPE='E')
```
Comment (JasonMWhite) [inline python/tests/test_dwelling.py]: this mutates the object `sample_input_d` which is potentially shared amongst other test fixtures for the same test. This is to be avoided.
```
output = sample_input_d.copy()
output['EVAL_TYPE'] = 'E'
return output
```

File: python/tests/test_dwelling.py [hunk 10]
```diff
@@ -105,6 +126,19 @@ def test_from_row(self, sample_input_d: reader.InputData) -> None:
...
             forward_sortation_area='K1P',
+            ceilings=[
+                dwelling.Ceiling(
+                    label='Main attic',
+                    type_english='Attic/gable',
+                    type_french='Combles/pignon',
+                    nominal_rsi=2.864,
+
+                    effective_rsi=2.9463,
+
```
Comment (JasonMWhite) [inline python/tests/test_dwelling.py]: 🔥

**Pull request merged**